vmdk_plugin/drivers/vmdk/vmdk_driver.go

```diff
@@ -190,6 +192,80 @@ func (d *VolumeDriver) UnmountVolume(name string) error {
 	return d.ops.Detach(name, nil)
 }
 
+// private function that does the job of mounting volume in conjunction with refcounting
+// Called under a lock if refcounting hasn't been successful, else without a lock
+// Don't do refcounting, just the mount if refcounting hasn't been successful.
+func (d *VolumeDriver) processMount(r volume.MountRequest, doRefcnt bool) volume.Response {
+	// If the volume is already mounted , just increase the refcount.
+	//
+	// Note: We are deliberately incrementing refcount first, before trying
+	// to do anything else. If Mount fails, Docker will send Unmount request,
+	// and we will happily decrement the refcount there, and will fail the unmount
+	// since the volume will have been never mounted.
+	// Note: for new keys, GO maps return zero value, so no need for if_exists.
+
+	if doRefcnt {
+		refcnt := d.incrRefCount(r.Name) // save map traversal
+		log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
+		if refcnt > 1 {
+			log.WithFields(
+				log.Fields{"name": r.Name, "refcount": refcnt},
+			).Info("Already mounted, skipping mount. ")
+			return volume.Response{Mountpoint: getMountPoint(r.Name)}
+		}
+	}
+
+	// This is the first time we are asked to mount the volume, so comply
+	status, err := d.ops.Get(r.Name)
```

Comment:
status should already include "datastore" which can be used as ref counting key and mount dir name. I am okay if you address mount/unmount issue in separate PR